### PR TITLE
fix(rgbw): NNLS out-of-hull projection (#2708) + in-test ground truth

### LINKED ISSUE #2708: bug(rgbw): out-of-hull source colors return (0,0,0,0) — missing NNLS projection (gist §3)
## Summary

When a source RGB input maps to a target chromaticity that falls **outside the LED's 4-emitter convex hull**, the strict sub-gamut solver returns `(0, 0, 0, 0)` instead of projecting to the nearest in-hull point. The math-model gist's §3 specifies a constrained NNLS projection (\"Builder and verifier must use the same projection — this is a correctness requirement\") that FastLED has not yet implemented.

Surfaced by a side-by-side comparison against the Python reference (https://gist.github.com/JChalka/cebc5a018066cae05d98cf9088c3b3b9):

| Input         | FastLED            | Reference          | Δ                  |
|---------------|--------------------|--------------------|--------------------|
| sRGB pure blue| `(0, 0, 0, 0)`     | `(26, 0, 242, 0)`  | (+26, 0, +242, 0)  |

All other in-hull test cases (D65 white, red, green, cyan, magenta, yellow, mid grey, etc.) match the reference within ±16/255 — only out-of-hull cases black out.

## Root cause

`solve_strict_subgamut` in `src/fl/gfx/rgbw_colorimetric.cpp.hpp` iterates the three sub-gamut triangles (RGW / RBW / BGW). When **none** of them contain the target xy (barycentric check at lines 105-112), the function returns `false`, and the dispatch in `rgbw.cpp.hpp:303-309` falls back to `rgb_2_rgbw_exact` — which itself can't represent a chromaticity outside the LED hull and zeroes the W contribution incorrectly. Net result for the sRGB-blue case is `(0,0,0,0)`.

The reference, by contrast, runs `_strict_project_target_xyz_to_led_hull(target_xyz)` first (lines 2148-… of `xy_target_rgbw_model.py`), which uses non-negative least squares (NNLS) to find the closest in-hull XYZ before routing into a sub-gamut.

## Why this matters

1. **Standard color spaces routinely escape the LED hull.** sRGB's blue primary at xy(0.15, 0.06) is *barely* outside the SK6812-class LED's RBW triangle. Anything saturated and short-wavelength (pure blue, deep violet) blacks out.
2. **Verification fails on these inputs.** The user's D65-targeted verification tool reports `(0,0,0,0)` from FastLED vs. real RGBW values from the spec, masking the actual chromaticity reproduction quality of everything else.
3. **The fallback is wrong.** `rgb_2_rgbw_exact` was designed for in-gamut inputs — for an out-of-hull target it produces a chromaticity excursion that's worse than the NNLS projection.

## Proposed fix

1. Implement constrained NNLS in `colorimetric_detail` (small enough to inline; 4-variable non-negative quadratic with 3 equality constraints). Apply to `X_t` before triangle routing if a target chromaticity is flagged out-of-hull by an initial barycentric scan.
2. Add a test case asserting sRGB pure blue produces a non-zero RGBW with reconstructed chromaticity close to the LED hull's blue boundary (not the original sRGB blue).
3. The boosted (`solve_wx_overdrive`) path needs the same projection — currently it tolerates out-of-hull targets via residual clamping, but the chromaticity drift is uncontrolled.

## References

- Parent: #2545
- Implementing PR for input color space + boosted differentiation: #2707
- Math model spec: https://gist.github.com/JChalka/915603d7930b7216d5c5aae834ec2ea9 (§3)
- Reference implementation: https://gist.github.com/JChalka/cebc5a018066cae05d98cf9088c3b3b9 (`_strict_project_target_xyz_to_led_hull`)
- Comparison harness producing the above table lives in `compare_rgbw.py` (uncommitted) — fetches the gist, drives FastLED's algorithm (faithfully mirrored in Python), and prints the side-by-side delta.

### PR DESCRIPTION
**Pull request by zackees**
Closes #2708.

Single PR that lands both halves of the math-model gist's §3 + §5 contract:

1. **Implements** the gist §3 NNLS out-of-hull projection in production code so sRGB-blue / Rec.2020-green inputs no longer collapse to (0,0,0,0).
2. **Ports** the reference's strict sub-gamut + projection algorithm into the unit test as an independent ground-truth source, so future regressions in either the production code or the test mirror surface immediately.

## #2708 — out-of-hull NNLS projection

When a source RGB triple mapped to an XYZ target outside the LED 4-emitter convex hull, the strict solver returned (0,0,0,0) because no sub-gamut triangle accepted the chromaticity. sRGB blue at xy ≈ (0.15, 0.06) was the canonical case — outside the LED's RBW and BGW triangles on the default profile.

**Implementation**:

- `colorimetric_detail::nnls3` (inline in `rgbw_colorimetric.h`): projected-gradient NNLS for 3×3 systems (500 iters @ step 0.01), matching the reference's scipy-less fallback. Alloc-free; only invoked when the target is out-of-hull.
- `project_to_hull` (static in `rgbw_colorimetric.cpp.hpp`): runs NNLS in each of the three sub-gamuts (RGW / RBW / BGW), caps drive at 1.0 per sub-gamut, selects min-residual. Returns the achieved chromaticity at unit Y.
- Wired into `solve_strict_subgamut`, `solve_strict_subgamut_xy`, and `solve_wx_overdrive`. The LUT builder reuses `solve_strict_subgamut_xy` so the rebuilt LUT cells contain projected RGBW for boundary samples.

## In-test ground truth (no Python at test time)

`tests/fl/gfx/rgbw_colorimetric.cpp` gets two namespaces:

- **`reference::`** — port of `xy_target_rgbw_model.py` strict sub-gamut + §3 NNLS projection. Independent of production code.
- **`fastled_mirror::`** — text-copy of the production solver (kept in sync with `.cpp.hpp` for tests that can't link the colorimetric path).

### Test cases

1. **`FastLED strict subgamut agrees with reference port (in-hull)`** — 10 inputs (D65, primaries, secondaries, mixed). Asserts every channel within ±24 / 255 of the reference.
2. **`out-of-hull sRGB blue: FastLED projects to match reference (#2708)`** — was the executable contract for #2708; now asserts FastLED produces non-trivial output AND matches the reference within ±1 / 255 (rounding only — both share the same NNLS).
3. **plus the existing source-space / boosted / Hermite tests from #2707.**

## Verification trail

Before trusting the C++ port as ground truth, three independent implementations were checked against each other on 14 test inputs across strict / boosted / native-gamut modes:

1. **Actual FastLED solver** (production `.cpp.hpp`) — driven by a standalone harness compiled directly with clang++.
2. **C++ mirror in this test TU** (`fastled_mirror::strict_subgamut`) — text-copy of the production code.
3. **Python mirror** (`compare_rgbw.py` — scratch file, fetches the gist, mirrors FastLED's algorithm in Python).

All three produced bit-identical RGBW across every case before #2708 was implemented. After #2708, the reference port and FastLED now also agree on out-of-hull cases.

## Drive-by fixes

- `src/platforms/arm/teensy/audio_input_teensy_config.h` gains `// IWYU pragma: private`.
- `ci/lint_cpp/rust_bridge.py` drops a redundant `except KeyboardInterrupt: raise` that the KBI checker flagged. Plain re-raise was a no-op.

## Test plan

- [x] `bash lint` clean (Python, C++, JS)
- [x] `bash test --cpp --clean` passes 308/308 including the new ground-truth cases
- [x] sRGB pure blue now produces non-trivial RGBW matching the reference
- [ ] CI green
- [ ] CodeRabbit review addressed

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added a Python RGBW comparison tool to run FastLED vs. reference colorimetric models and diagnostics.
  * Improved RGBW solver behavior to project out-of-gamut targets onto the achievable gamut (preserves non-zero, chromatically consistent outputs).

* **Tests**
  * Added ground-truth regression tests validating strict-subgamut and projected solutions against the reference model, including an out-of-gamut blue case.

* **Chores**
  * Minor lint/config header annotation update.

### CODE REVIEW

File: compare_rgbw.py [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+    ("teal", 0, 200, 180),
+]
+
+
+def fmt(rgbw):
+    return "(%3d,%3d,%3d,%3d)" % rgbw
+
+
+def delta(a, b):
+    return tuple(int(b[i]) - int(a[i]) for i in range(4))
```
[the remark this answers is not in the document]
Comment (zackees) [reply compare_rgbw.py:264]: Moot — compare_rgbw.py was removed in eb2694d (it was a one-off scratch comparison file, not production code).

File: compare_rgbw.py [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+
+def main():
+    print(
+        "Profile: %d×%d×%d × %d-LED primaries from reference; source = Rec709/D65 (linear)"
+        % (1, 1, 1, 4)
+    )
+    print()
+    print(
+        "Test case          | %-18s | %-18s | %-22s"
+        % ("FastLED strict", "Reference sub_gamut", "FastLED − Ref (∆R,G,B,W)")
```
[the remark this answers is not in the document]
Comment (zackees) [reply compare_rgbw.py:275]: Moot — compare_rgbw.py was removed in eb2694d (it was a one-off scratch comparison file, not production code).

File: compare_rgbw.py [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+
+def invert3x3(m: np.ndarray) -> tuple[np.ndarray, bool]:
+    """Mirror invert3x3 — returns (inv, ok)."""
+    det = np.linalg.det(m)
+    if abs(det) < 1e-20:
+        return np.zeros((3, 3)), False
+    return np.linalg.inv(m), True
+
+
+def build_source_matrix(xy_r, xy_g, xy_b, xy_w) -> tuple[np.ndarray, bool]:
```
[the remark this answers is not in the document]
Comment (zackees) [reply compare_rgbw.py:48]: Moot — compare_rgbw.py was removed in eb2694d (it was a one-off scratch comparison file, not production code).

File: compare_rgbw.py [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+        self.P_RBW_inv, _ = invert3x3(P_RBW)
+        self.P_BGW_inv, _ = invert3x3(P_BGW)
+        self.d_W = self.P_RGB_inv @ self.P_W
+        # Source space (#2705)
+        self.M_src, self.has_source_space = build_source_matrix(
+            profile["input_xy_r"],
+            profile["input_xy_g"],
+            profile["input_xy_b"],
+            profile["input_xy_w"],
+        )
```
[the remark this answers is not in the document]
Comment (zackees) [reply compare_rgbw.py:106]: Moot — compare_rgbw.py was removed in eb2694d (it was a one-off scratch comparison file, not production code).

File: tests/fl/gfx/rgbw_colorimetric.cpp [hunk 5]
```diff
@@ -484,3 +484,368 @@ FL_TEST_CASE("LUT stride constants match interp scheme") {
...
+    pack(cache->P_B, cache->P_G, cache->P_W, P_BGW);
+    invert3x3(P_RGB, cache->P_RGB_inv);
+    invert3x3(P_RGW, cache->P_RGW_inv);
+    invert3x3(P_RBW, cache->P_RBW_inv);
+    invert3x3(P_BGW, cache->P_BGW_inv);
+    matvec3(cache->P_RGB_inv, cache->P_W, cache->d_W);
+    cache->has_source_space = (p->input_xy_w[1] > 1e-6f) &&
+        build_source_matrix(p->input_xy_r, p->input_xy_g, p->input_xy_b,
+                            p->input_xy_w, cache->M_src);
+}
```
[the remark this answers is not in the document]
Comment (zackees) [reply tests/fl/gfx/rgbw_colorimetric.cpp:544]: Fixed in eb2694d — mirror now zero-inits M_src on build_source_matrix failure to match production.
